### PR TITLE
[new release] ocaml-r (0.6.0)

### PR DESCRIPTION
**Pull request by pveber**
Objective Caml bindings for the R interpreter

- Project page: <a href="https://github.com/pveber/ocaml-r/">https://github.com/pveber/ocaml-r/</a>
- Documentation: <a href="https://pveber.github.io/ocaml-r/ocaml-r/index.html">https://pveber.github.io/ocaml-r/ocaml-r/index.html</a>

##### CHANGES:

- update wrt to R 4.2, which restricts access to R internals. As a
  consequence, many functions have been removed. Stubgen is not
  possible anymore, and many wrappers have been rewritten by hand, so
  that high level bindings are not affected in the end.
- added table and qqplot

### CODE REVIEW

File: packages/ocaml-r/ocaml-r.0.6.0/opam [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  "alcotest" {with-test}
+  "base" {build & >= "v0.14"}
+  "conf-r" {build}
+  "conf-r-mathlib" {build}
+  "dune" {>= "2.5"}
+  "ocaml" {>= "4.08"}
+  "stdio" {>= "v0.14" & build}
+]
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/ocaml-r/ocaml-r.0.6.0/opam:26]:
```suggestion
  ["dune" "subst"] {dev}
```

**Comment by mseri** [comment]
It is failing on macos only on 4.14 with
```
#=== ERROR while compiling ocaml-r.0.6.0 ======================================#
# context              2.1.4 | macos/arm64 | ocaml-base-compiler.4.14.1 | pinned(https://github.com/pveber/ocaml-r/releases/download/v0.6.0/ocaml-r-0.6.0.tbz)
# path                 ~/.opam/4.14.1/.opam-switch/build/ocaml-r.0.6.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ocaml-r -j 7 @install
# exit-code            1
# env-file             ~/.opam/log/ocaml-r-16547-b8925a.env
# output-file          ~/.opam/log/ocaml-r-16547-b8925a.out
### output ###
# File "lib/dune", line 1, characters 0-671:
#  1 | (library
#  2 |   (name             OCamlR)
#  3 |   (public_name      ocaml-r)
# ....
# 19 |                     reduction_stub)
# 20 |     (flags         (:include c_flags.sexp)))
# 21 |   (c_library_flags (:include c_library_flags.sexp)))
# (cd _build/default && /Users/mac1000/.opam/4.14.1/bin/ocamlmklib.opt -g -o lib/OCamlR_stubs lib/write_internal_stub.o lib/symbols_stub.o lib/sexp.o lib/s4_stub.o lib/s3_stub.o lib/reduction_stub.o lib/read_internal_stub.o lib/r_parser_stub.o lib/initialisation_stub.o lib/databridge.o lib/conversion_stub.o lib/allocation_stub.o -ldopt -L/opt/homebrew/opt/gettext/lib -ldopt -L/opt/homebrew/opt/readline/lib -ldopt -L/opt/homebrew/opt/xz/lib -ldopt -L/opt/homebrew/Cellar/r/4.2.2_1/lib/R/lib -ldopt -L/opt/homebrew/Cellar/r/4.2.2_1/lib -ldopt -lR -ldopt -lRmath)
# ld: file not found: @rpath/libquadmath.0.dylib for architecture arm64
# clang: error: linker command failed with exit code 1 (use -v to see invocation)
# File "lib/math/dune", line 1, characters 0-270:
# 1 | (library
# 2 |   (name            OCamlR_math)
# 3 |   (public_name     ocaml-r.math)
# 4 |   (wrapped         false)
# 5 |   (foreign_stubs
# 6 |     (language c)
# 7 |     (names         r_math_stubs)
# 8 |     (flags         (:include ../c_flags.sexp)))
# 9 |   (c_library_flags (:include ../c_library_flags.sexp)))
# (cd _build/default && /Users/mac1000/.opam/4.14.1/bin/ocamlmklib.opt -g -o lib/math/OCamlR_math_stubs lib/math/r_math_stubs.o -ldopt -L/opt/homebrew/opt/gettext/lib -ldopt -L/opt/homebrew/opt/readline/lib -ldopt -L/opt/homebrew/opt/xz/lib -ldopt -L/opt/homebrew/Cellar/r/4.2.2_1/lib/R/lib -ldopt -L/opt/homebrew/Cellar/r/4.2.2_1/lib -ldopt -lR -ldopt -lRmath)
# ld: file not found: @rpath/libquadmath.0.dylib for architecture arm64
# clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
and
```

#=== ERROR while compiling ocaml-r.0.6.0 ======================================#
# context              2.1.4 | macos/x86_64 | ocaml-base-compiler.4.14.1 | pinned(https://github.com/pveber/ocaml-r/releases/download/v0.6.0/ocaml-r-0.6.0.tbz)
# path                 ~/.opam/4.14.1/.opam-switch/build/ocaml-r.0.6.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ocaml-r -j 11 @install
# exit-code            1
# env-file             ~/.opam/log/ocaml-r-63570-fd96a8.env
# output-file          ~/.opam/log/ocaml-r-63570-fd96a8.out
### output ###
# File "lib/dune", line 1, characters 0-671:
#  1 | (library
#  2 |   (name             OCamlR)
#  3 |   (public_name      ocaml-r)
# ....
# 19 |                     reduction_stub)
# 20 |     (flags         (:include c_flags.sexp)))
# 21 |   (c_library_flags (:include c_library_flags.sexp)))
# (cd _build/default && /Users/mac1000/.opam/4.14.1/bin/ocamlmklib.opt -g -o lib/OCamlR_stubs lib/write_internal_stub.o lib/symbols_stub.o lib/sexp.o lib/s4_stub.o lib/s3_stub.o lib/reduction_stub.o lib/read_internal_stub.o lib/r_parser_stub.o lib/initialisation_stub.o lib/databridge.o lib/conversion_stub.o lib/allocation_stub.o -ldopt -L/usr/local/opt/gettext/lib -ldopt -L/usr/local/opt/readline/lib -ldopt -L/usr/local/opt/xz/lib -ldopt -L/usr/local/Cellar/r/4.2.2/lib/R/lib -ldopt -L/usr/local/Cellar/r/4.2.2/lib -ldopt -lR -ldopt -lRmath)
# ld: file not found: @rpath/libquadmath.0.dylib for architecture x86_64
# clang: error: linker command failed with exit code 1 (use -v to see invocation)
# File "lib/math/dune", line 1, characters 0-270:
# 1 | (library
# 2 |   (name            OCamlR_math)
# 3 |   (public_name     ocaml-r.math)
# 4 |   (wrapped         false)
# 5 |   (foreign_stubs
# 6 |     (language c)
# 7 |     (names         r_math_stubs)
# 8 |     (flags         (:include ../c_flags.sexp)))
# 9 |   (c_library_flags (:include ../c_library_flags.sexp)))
# (cd _build/default && /Users/mac1000/.opam/4.14.1/bin/ocamlmklib.opt -g -o lib/math/OCamlR_math_stubs lib/math/r_math_stubs.o -ldopt -L/usr/local/opt/gettext/lib -ldopt -L/usr/local/opt/readline/lib -ldopt -L/usr/local/opt/xz/lib -ldopt -L/usr/local/Cellar/r/4.2.2/lib/R/lib -ldopt -L/usr/local/Cellar/r/4.2.2/lib -ldopt -lR -ldopt -lRmath)
# ld: file not found: @rpath/libquadmath.0.dylib for architecture x86_64
# clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
Do you know why it can be the case?

**Comment by pveber** [comment]
Thanks @mseri! It is puzzling indeed! Are the macos images the same for 4.14 and 5.0? If they're different, we can assume this is a (homebrew) packaging bug, right?

**Comment by mseri** [comment]
Or an ocaml 4.14 one? It is pizzling that onmacos with ocaml 5 it all goes well. I don’t know. I will try it on my machines with macports, just to compare, but it will take a couple of days

**Comment by mseri** [comment]
Works fine locally. By the way, on macports to make it work you need
```
diff --git a/lib/config/discover.ml b/lib/config/discover.ml
index 9425433..9a8ca6c 100644
--- a/lib/config/discover.ml
+++ b/lib/config/discover.ml
@@ -9,8 +9,8 @@ let cfig c =
       libs = ["-lR"; "-lRmath"];
       cflags = [] }
   in
-  let rlib = C.Process.run c "pkg-config" ["--libs-only-L"; "--libs-only-l"; "libR"; "libRmath"] in
-  let rflg = C.Process.run c "pkg-config" ["--cflags"; "libR"; "libRmath"] in
+  let rlib = C.Process.run c "pkg-config" ["--libs-only-L"; "--libs-only-l"; "libR"] in
+  let rflg = C.Process.run c "pkg-config" ["--cflags"; "libR"] in
   let chk = (rlib.exit_code, rflg.exit_code) in
   let lib = trim_to_list rlib.stdout in
   let flg = trim_to_list rflg.stdout in
```
since there is no `libRmath.pc`. I would suggest separating the two calls for `libR` and `libRmath`. In this way it will get the correct flags (if available) from the first call and survive the failure of the second.

**Comment by pveber** [comment]
Thanks a lot for the suggestion @mseri! It's upstreamed in pveber/ocaml-r@aa96dc5 .

**Comment by ulrikstrid** [comment]
I wonder if this is the same failure we've seen on nix

**Comment by pveber** [comment]
Is there a log that can be checked somewhere?

**Comment by ulrikstrid** [comment]
@pveber it's the same, can't find `R.h`. Your patch worked so all good! :+1:

**Pull request merged**